### PR TITLE
#167729793 Users should book a return Trip

### PR DESCRIPTION
**Pull request by davidfrank96**
#### What does this PR do?
** A User will be able to book a return Trip 
#### Description of Task to be completed?
- created a return Trip controller
- Created a Trips table
- created a validation Handler for handling errors
- created a date validation function
- created not empty function to validate the req.body
- created an Authorization controller to validate the user token
#### How should this be manually tested?
** npm pull from the origin Branch from staging
#### Any background context you want to provide?
N/A
#### What are the relevant pivotal tracker stories?
[Finishes #167729793] 
#### Screenshots (if appropriate)
ReturnTrip

#### Questions:
N/A

### CODE REVIEW

File: src/controllers/returnTrip.js [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+                departureDate, returnDate, currentOfficeLocation, accomodation, reason, tripType, destination
+            } = req.body;
+            const departureDateUTC = new Date(departureDate);
+            const returnDateUTC = new Date(returnDate);
+
+            const duration = validatedate(departureDateUTC, returnDateUTC);
+            if (!duration) {
+                return res.status(400).json({
+                    status: 'error',
+                    message: 'Departure date and Return date cant be the same'
```
Comment (chibytez) [inline src/controllers/returnTrip.js]: Can you please fix your can't

File: src/controllers/returnTrip.js [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+                });
+            }
+
+            const tripsResult = await Trips.create({
+                currentOfficeLocation,
+                tripId: uuidv4(),
+                user_id: id,
```
Comment (chibytez) [inline src/controllers/returnTrip.js]: We are using camel case not snake case

File: src/controllers/returnTrip.js [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+                reason,
+                tripType,
+                requestStatus: 'pending',
+                destination
+            });
+            if (tripsResult) {
+                return res.status(201).json({
+                    status: 'success',
+                    data: {
+                        user_id: id,
```
Comment (chibytez) [inline src/controllers/returnTrip.js]: Deto

File: src/controllers/returnTrip.js [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+                        destination,
+                        current_office_location: currentOfficeLocation,
+                        departure_date: departureDateUTC,
+                        return_date: returnDateUTC,
```
Comment (chibytez) [inline src/controllers/returnTrip.js]: deto

File: src/database/migrations/20190901024921-create-Trips.js [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+      allowNull: false,
+      autoIncrement: true,
+      type: Sequelize.INTEGER,
+      primaryKey: true
+    },
+    tripId: {
+      type: Sequelize.STRING,
+      unique: true
+    },
+    user_id: {
```
Comment (chibytez) [inline src/database/migrations/20190901024921-create-Trips.js]: Can we please change this to camel case

File: src/test/returntrip.test.js [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+                user_id: '',
+                departureDate: new Date(departureDate).toUTCString(),
+                returnDate: new Date(returnDate).toUTCString(),
+                accomodation: '',
+                reason: '',
+                tripType: '',
+                requestStatus: 'pending',
+                destination: '',
+            })
+            .end((err, res) => {
```
Comment (chibytez) [inline src/test/returntrip.test.js:293]: Can we also take care of the err

**Review by chibytez — CHANGES_REQUESTED** [review]

**Review by chibytez — CHANGES_REQUESTED** [review]
Can we add a screenshot to show that the test is passing

File: src/controllers/returnTrip.js [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+            const { id } = req;
+            const { departureDate, returnDate, currentOfficeLocation, accomodation, reason, tripType, destination} = req.body;
+            const departureDateUTC = new Date(departureDate);
+            const returnDateUTC = new Date(returnDate);
+
+            const duration = validatedate(departureDateUTC, returnDateUTC);
+            if (!duration) {
+                return res.status(400).json({
+                    status: 'error',
+                    message: 'Departure and Return date can"t be the same'
```
Comment (chibytez) [inline src/controllers/returnTrip.js]: can we fix it to can't

File: src/models/Trip.js [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+    const Trips = sequelize.define('Trips', {
+        tripId: {
+            type: DataTypes.STRING,
+            required: true
+        },
+        user_id: {
+            type: DataTypes.INTEGER,
+            required: true
+        },
+        departureDate: {
```
Comment (sebaztyn) [inline src/models/Trip.js]: Based on @Stylll 's recommendation, this should be a date type

File: src/models/Trip.js [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+            type: DataTypes.STRING,
+            required: true
+        },
+        returnDate: {
```
Comment (sebaztyn) [inline src/models/Trip.js]: This should have a date type too

File: src/models/Trip.js [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+        },
+        accomodation: {
+            type: DataTypes.TEXT,
+            required: true
+        },
+        requestStatus: {
+            type: DataTypes.STRING,
+            required: true
+        },
+        destination: {
```
Comment (sebaztyn) [inline src/models/Trip.js]: We are advised to make this an array

File: src/routes/api/index.js [hunk 11]
```diff
@@ -11,6 +12,7 @@ usersRouter.use('/v1/users', passwordRouter);
 usersRouter.get('/', (req, res) => res.status(200).send('Welcome to  Shadowcat API'));
 
 usersRouter.use('/v1/users', passwordRouter);
+usersRouter.use('/v1/trips', returntrip);
```
Comment (chibytez) [inline src/routes/api/index.js]: I think it should be /v1/users/trips

Comment (sebaztyn) [inline src/routes/api/index.js]: All trip requests for oneway, return and multi-city should use just one route based on the advice of @Stylll

File: src/routes/api/returntrip.js [hunk 12]
```diff
@@ -0,0 +1,12 @@
+import express from 'express';
+import ReturnTrip from '../../controllers/returnTrip';
+import returntripvalidate from '../../validation/returntripvalidate';
+import ValidationHandler from '../../utils/ValidationHandler';
+import Auth from '../../middlewares/auth';
+
+
+const router = express.Router();
+
+router.post('/request', Auth.authenticate, returntripvalidate.returnTrip, ValidationHandler.validate,  ReturnTrip.returnTripRequest );
```
Comment (chibytez) [inline src/routes/api/returntrip.js]: I think your validation should be the first route

File: src/utils/ValidationHandler.js [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+     */
+    static validate(req, res, next) {
+        const errors = validationResult(req);
+        req = { ...req, ...matchedData(req) };
+
+        if (!errors.isEmpty()) {
+
+            const mappedErrors = errors.mapped();
+
+            let results = Object.keys(mappedErrors).map(function (key) {
```
Comment (chibytez) [inline src/utils/ValidationHandler.js]: Can we use arrow function here

File: src/utils/helper/notEmpty.js [hunk 14]
```diff
@@ -0,0 +1,19 @@
+/**
+ * Function to check if value is empty
+ * @param {string} value
+ * @param {string} msg
+ * @return {(error|bool)} returns error or true
+ */
+function notEmpty(value, msg) {
```
Comment (chibytez) [inline src/utils/helper/notEmpty.js]: deto

File: src/test/returntrip.test.js [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+                currentOfficeLocation: 'abuja',
+                tripId: uuidv4(),
+                userId: '567',
+                departureDate: '2019-09-12',
+                returnDate: '2019-09-26',
+                accomodation: 'transcop',
+                reason: 'Holiday',
+                tripType: 'Return-trip',
+                requestStatus: 'pending',
+                destination: 'Lagos',
```
Comment (sebaztyn) [inline src/test/returntrip.test.js]: Destination should be array of integers and current location should be an integer representing the current office location

**Review by chibytez — CHANGES_REQUESTED** [review]

**Review by sebaztyn — CHANGES_REQUESTED** [review]

**Pull request closed without merging**